### PR TITLE
Blog App - processing data in models

### PR DESCRIPTION
**Pull request by AmaliaGM**
# Project requirements

- Create models
Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
Set up associations between models.
- Use models to insert data
Familiarize yourself with ['rails c' command](https://guides.rubyonrails.org/command_line.html#bin-rails-console).
Open rails console for your project.
Create at least one user by running the following code:
   first_user = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
   second_user = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')
Create at least 4 posts written by one of the users you created by running the following code:
   first_post = Post.create(author: first_user, title: 'Hello', text: 'This is my first post')
Create at least 6 posts comments for one of the posts you created by running the following code:
   Comment.create(post: first_post, author: second_user, text: 'Hi Tom!' )
Use other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.
- Create custom methods
Your models should include some additional methods.

    - Users
    A method that returns the 3 most recent posts for a given user.
    - Posts
    A method that updates the posts counter for a user.
    A method which returns the 5 most recent comments for a given post.
    - Comments
    A method that updates the comments counter for a post.
    - Likes
    A method that updates the likes counter for a post.
    - Go to rails c and check if your methods are working.

### CODE REVIEW

File: app/models/comment.rb [hunk 1]
```diff
@@ -1,4 +1,4 @@
 class Comment < ApplicationRecord
-  belongs_to :author, class_name: 'User'
-  belongs_to :post, counter_cache: :comments_counter
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
+  belongs_to :post, counter_cache: :comments_counter, foreign_key: :post_id
```
Comment (VICTOR-CHUKWU) [inline app/models/comment.rb:3]: - [ ] comment should have a method that updates comments counter
```def update_comments_counter
  post.increment!(:comments_counter)
end
```

File: app/models/like.rb [hunk 2]
```diff
@@ -1,4 +1,4 @@
 class Like < ApplicationRecord
-  belongs_to :author, class_name: 'User'
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
   belongs_to :post, counter_cache: :likes_counter
```
Comment (VICTOR-CHUKWU) [inline app/models/like.rb:3]: - [ ] Likes should also have a method that updates like counter

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @AmaliaGM  👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good PR
- Good commit
- All linters passed


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/comment.rb (hunk 1 above)
Comment (AmaliaGM) [reply app/models/comment.rb:3]: @VICTOR-CHUKWU  Thanks for your feed back!

File: app/models/like.rb (hunk 2 above)
Comment (AmaliaGM) [reply app/models/like.rb:3]: @VICTOR-CHUKWU  it does have a counter and comments does also. Likes does not have likes that's why it's nill, but if you Post.all you'll see

File: app/models/comment.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
...
+  post.increment!(:comments_counter)
```
Comment (codepantha) [inline app/models/comment.rb]: - [ ] Kindly create a method and place the highlighted piece of code inside it. You could name it `update_comments_counter` if you like. This is what the previous reviewer meant 👍

File: app/models/like.rb [hunk 4]
```diff
@@ -1,4 +1,4 @@
 class Like < ApplicationRecord
-  belongs_to :author, class_name: 'User'
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (codepantha) [inline app/models/like.rb:2]: - [ ] Just like the previous reviewer suggested, please create a method that updates the likes counter. The resultant code will look something like this 👇 
``` ruby
def update_likes_counter
  post.increment!(:likes_counter)
end
```

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: REQUIRES CHANGES ♻️ 

## Hi @AmaliaGM 👋, 

Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Highlights 🎯 
✔️ Descriptive PR
✔️ Good commit history
✔️ No linter errors

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

- [ ] I strongly recommend that you create a professional README for your project using this new [template](https://github.com/microverseinc/readme-template). Your current README is not as professional as it should be and your submission should have been invalidated, but since the previous reviewer missed it, I'm making it an optional suggestion. Kindly make this change 🙏 


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/comment.rb (hunk 3 above)
Comment (AmaliaGM) [reply app/models/comment.rb]: @codepantha Hi. The comments are being counted
image

**Comment by medaminedev66** [comment]
Hello @AmaliaGM  👋,

_This is an additional review as a result of the second opinion request._

Following the project requirements, you should create a method that updates the likes/comments counter for a post:
image

To learn more about how methods in ruby, I recommend checking out[ this link](https://www.tutorialspoint.com/ruby/ruby_methods.htm).
 
Great job incrementing the comments counter, but it should be inside a method as shown below (the same thing for the likes counter in the Like model):
```rb
def update_comments_counter
  post.increment!(:comments_counter)
end
```

Keep rocking 🚀!

File: app/models/like.rb [hunk 5]
```diff
@@ -1,4 +1,7 @@
 class Like < ApplicationRecord
   belongs_to :author, class_name: 'User', foreign_key: :author_id
   belongs_to :post, counter_cache: :likes_counter
+  def update_likes_counter
+    likes.increment!(:likes_counter)
+  end
```
Comment (codepantha) [inline app/models/like.rb]: - [ ] You have done well by creating the `update_likes_counter` method. However, in order to increment the value in the `likes_counter` column which is in the `posts` table, you need to take advantage of the relationship.
I suggest you do it this way 👇
``` ruby
def update_likes_counter
  post.increment!(:likes_counter)
end
```

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: REQUIRES CHANGES ♻️ 

## Hi @AmaliaGM 👋, 

Great work on making the changes requested by the previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Highlights 🎯 
✔️ Descriptive PR
✔️ Good commit history
✔️ No linter errors

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by omar-muhamad — APPROVED** [review]
# Approved ✔
### Hi @AmaliaGM,

Great job, Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

thumbs-up-agree


## (Highlights) Good Points: :+1:
- You documented your work professionally. ✔
- You used the correct workflow.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- N/A

## [Optional] suggestions:

- N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**